Enter editing mode via Enter or Spacebar

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Allow entering into edit mode with keyboard enter/spacebar press on the site editor frame in view mode.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
There was a regression in being able to enter editing mode via keydown on the site editor due to the onKeydown event being blocked. 

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This allows an onKeydown prop to be fired if passed to the iframe. I'm not positive this is the best way.

I also added a test to check for this as well as general keyboard navigation of the site editor in view mode.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
- Go to the site editor
- Use the keyboard to move focus to the site editor 
- Focus ring should be present on iframe
- Press Enter/Spacebard
- Should be in editing mode

jeryj committed:
Enter editing mode via Enter or Spacebar

There was a regression in being able to enter editing mode via keydown on the site editor due to the onKeydown event being blocked. This allows an onKeydown prop to be fired if passed to the iframe. I'm not positive this is the best way.

I also added a test to check for this as well as general keyboard navigation of the site editor in view mode.

scruffian commented on test/e2e/specs/site-editor/navigation.spec.js:
```diff
@@ -0,0 +1,80 @@
+		await admin.visitSiteEditor();
+
+		// Navigate to the iframe
+		await pageUtils.pressKeys( 'Tab', { times: 3 } );
```
This feels quite fragile. Is there a way to focus the iframe directly without relying on the number of tab presses?

draganescu replied:
I think this is correct because we should be aware we introduced a new tab stop?

scruffian commented on test/e2e/specs/site-editor/navigation.spec.js:
```diff
@@ -0,0 +1,80 @@
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+test.describe( 'Site editor navigation', () => {
```
This test feels quite big and I worry that it's going to need a lot of work to maintain it as the editor continues to evolve. I think it would be better if we could find a way to just test the spacebar press on its own.

draganescu commented on test/e2e/specs/site-editor/navigation.spec.js:
```diff
@@ -0,0 +1,80 @@
+		await requestUtils.activateTheme( 'twentytwentyone' );
+	} );
+
+	test( 'Can use keyboard to navigate the site editor', async ( {
```
'`Can use keyboard to navigate the site editor` sounds like the whole battery of tests instead of a single test - and looking at the many expectations it's kinda true.

draganescu reviewed (commented):
This is great, I think we should split the test into a few smaller ones because the big one ... is too big 😄

jeryj commented:
It’s important to know we can use the keyboard to navigate between modes, so if we split this into several tests, it’s going to repeat a lot of the same work across multiple tests, not be as robust, and take way longer to run. For e2e tests, one larger test is preferable than multiple shorter ones. 

That said, I really don’t like the 9 shift tabs to go back. I should replace that with a region shortcut. On the other hand, maybe it’s a UX issue that we require 9 tab stops to return to the previous view. 

On Fri, Feb 09, 2024 at 6:55 AM Andrei Draganescu < Andrei Draganescu ( Andrei Draganescu ***@***.***> ) > wrote:

> 
> 
> 
> 
> 
> 
> 
> ***@***.**** commented on this pull request.
> 
> 
> 
> This is great, I think we should split the test into a few smaller ones
> because the big one ... is too big 😄
> 
> 
> 
> In test/e2e/specs/site-editor/navigation.spec.js (
> https://github.com/WordPress/gutenberg/pull/58795#discussion_r1484280484 )
> :
> 
> > @@ -0,0 +1,80 @@ +/** + * WordPress dependencies + */ +const { test,
> expect } = require( ***@***.***/e2e-test-utils-playwright' ); +
> +test.describe( 'Site editor navigation', () => { + test.beforeAll( async
> ( { requestUtils } ) => { + await requestUtils.activateTheme( 'emptytheme'
> ); + } ); + + test.afterAll( async ( { requestUtils } ) => { + await
> requestUtils.activateTheme( 'twentytwentyone' ); + } ); + + test( 'Can use
> keyboard to navigate the site editor', async ( {
> 
> ' Can use keyboard to navigate the site editor sounds like the whole
> battery of tests instead of a single test - and looking at the many
> expectations it's kinda true.
> 
> 
> 
> In test/e2e/specs/site-editor/navigation.spec.js (
> https://github.com/WordPress/gutenberg/pull/58795#discussion_r1484284087 )
> :
> 
> > + await requestUtils.activateTheme( 'emptytheme' ); + } ); + +
> test.afterAll( async ( { requestUtils } ) => { + await
> requestUtils.activateTheme( 'twentytwentyone' ); + } ); + + test( 'Can use
> keyboard to navigate the site editor', async ( { + admin, + page, +
> pageUtils, + } ) => { + await admin.visitSiteEditor(); + + // Navigate to
> the iframe + await pageUtils.pressKeys( 'Tab', { times: 3 } );
> 
> I think this is correct because we should be aware we introduced a new tab
> stop?
> 
> 
> 
> —
> Reply to this email directly, view it on GitHub (
> https://github.com/WordPress/gutenberg/pull/58795#pullrequestreview-1872308957
> ) , or unsubscribe (
> https://github.com/notifications/unsubscribe-auth/AAHMHOCUYBXPBFHDJKFHCU3YSYME3AVCNFSM6AAAAABC6SDOPGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZSGMYDQOJVG4
> ).
> You are receiving this because you were assigned. Message ID: <WordPress/gutenberg/pull/58795/review/1872308957
> @ github. com>
> 
> 
>

t-hamano commented:
What do you think about limiting the key that triggers edit mode to only the Enter key?

In common browsers, the spacebar is used to scroll pages. If you have a mouse, you can operate the mouse wheel on the canvas to see the entire canvas without entering edit mode. If we use the spacebar as a trigger key, users who only use the keyboard won't be able to do the same thing.

alexstine commented:
@draganescu 

> What do you think about limiting the key that triggers edit mode to only the Enter key?

If the element trigger is a button, enter/space keys should activate it.

Thanks.

t-hamano commented:
>If the element trigger is a button, enter/space keys should activate it.

This iframe has `role="button"` applied. However, when I tested it with NVDA, it seems that when I focus on this iframe, it simply reads "section".

jeryj commented:
This is a pretty big regression, so we should sort it out soon. l'm looking at the failing tests. Does anyone have thoughts on [the fix here](https://github.com/WordPress/gutenberg/pull/58795/files#diff-ab17abd38e687edd56573694ae3017e105e96b319571c190e9d4705c1872bc9cR276-R278)?

jeryj commented:
Tests are passing locally after rebasing, so I'm force pushing to see if the tests will pass.

t-hamano reviewed (commented):
The changes regarding `onKeydown` seem to make sense to me, but the e2e tests seem a bit redundant to me as they try to cover many scenarios in one test.

For example, how about dividing the test by scenario and clarifying what to expect, as shown below.

```javascript
test.beforeEach( async ( { admin } ) => {
	await admin.visitSiteEditor();
} );

test.describe( 'Site editor keyboard navigation', () => {
	test( 'should open command palette via button press', async ( {} ) => {} );
	// test( 'should open command palette via keyboard shortcut', async ( {} ) => {} );
	// test( 'should not lose focus when switching screens', async ( {} ) => {} );
	test( 'should enter into editing mode via button press', async ( {} ) => {} );
	test( 'should not lose focus when returning to view mode', async ( {} ) => {} );
} );
```

However, I think the following scenario is unnecessary.

> should open command palette via keyboard shortcut

What we're hoping for this time is not whether the keyboard shortcuts will work correctly, but whether the tab stops will work as expected.

> should not lose focus when switching screens

I believe this is covered in [the Navigator component's unit test](https://github.com/WordPress/gutenberg/blob/15d5fbce69921a65de14d7ee31300b03a5bc8cbf/packages/components/src/navigator/test/index.tsx#L593).

jeryj commented:
My philosophy on e2e tests is that they should be a more realistic test of how the app is used. Also, having one large test with multiple steps is faster due to not needing to run the admin page setup each time (5s vs 10s). It is also more likely to catch odd edge cases this way. 

For example, you might be able to click the pages navigation item, have focus on the back button, then press it and return to the page. But, does everything work in tandem? Can you do multiple steps and not lose focus along the way? Sometimes things have unintended side effects, and I think e2e tests are great for catching this.

So, while this is a big test, _everything here should work_. My main concern is that the test will give false negatives if the amount of tab stops changes. I don't want to test the amount of tab stops between the items. **I want to test that you can navigate the site editor with the keyboard without losing focus, and splitting this longer test into more tests will give me less confidence that I can do that.**

I will update the e2e test though to be clearer about what is being tested in each step, but leave them within one larger test. This is a pretty common pattern in our e2e testing, such as in [the navigation block frontend interactivity tests](https://github.com/WordPress/gutenberg/blob/fix/keydown-on-site-iframe/test/e2e/specs/editor/blocks/navigation-frontend-interactivity.spec.js#L165).

> I believe this is covered in [the Navigator component's unit test](https://github.com/WordPress/gutenberg/blob/15d5fbce69921a65de14d7ee31300b03a5bc8cbf/packages/components/src/navigator/test/index.tsx#L593).

I think that test is a more isolated example, not testing within the site editor.

jeryj committed:
Document tests. Solidify the iframe check. Remove flaky shift+tab from editor to sidebar navigation

jeryj committed:
Remove command palette focus tests

jeryj committed:
Make tabbing long distances less flakey

jeryj committed:
Extract tabTo logic to util

jeryj commented:
I made several changes to the e2e tests based on feedback:
- Removed the command palette tests
- Labelled each test section more clearly. It is still within one test item, but each one has its own test comment to encapsulate what it is doing more clearly.
- Moved the "long distance tabbing" by a specific number to a for loop with a limit so we aren't testing for number of tab stops, but if we can tab to the item.

jeryj committed:
Don't return anything from tabToX utils

t-hamano commented on test/e2e/specs/site-editor/navigation.spec.js:
```diff
@@ -0,0 +1,128 @@
+		// But did it do anything?
+		// Test to make sure a Tab keypress works as expected.
+		// As of this writing, we are in select mode and a tab
+		/// keypress will reveal the header template select mode
```
```suggestion
		// keypress will reveal the header template select mode
```

t-hamano commented on test/e2e/specs/site-editor/navigation.spec.js:
```diff
@@ -0,0 +1,128 @@
+
+		// Test: Can navigate into the iframe using the keyboard
+		await editorNavigationUtils.tabToNode( 'IFRAME', { times: 10 } );
+		// Getting the actual iframe as a cleaner locator was suprisingly tricky,
+		// so we're using a css selector with .is-focused which should be present when the iframe has focus.
+		await expect(
+			page.locator( 'iframe[name="editor-canvas"].is-focused' )
+		).toBeFocused();
```
```suggestion
		await expect(
			page.getByRole( 'button', { name: 'Editor Canvas' } )
		).toBeFocused();
```

The role of the editor canvas iframe in view mode should be a button, so how about changing it like this?

jeryj replied:
Great idea. That's much better. Thanks!

t-hamano reviewed (commented):
> Does anyone have thoughts on [the fix here](https://github.com/WordPress/gutenberg/pull/58795/files#diff-ab17abd38e687edd56573694ae3017e105e96b319571c190e9d4705c1872bc9cR276-R278)? 

This fix should make sense. Originally, Iframe components received most props as is, including `onKeyDown`.

However, #54565 appears to have overridden the `onKeyDown` prop.

cc @ellatrix

jeryj committed:
Remove extra forward slash

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jeryj committed:
Find editor canvas by button role

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jeryj committed:
Remove tabToNode. Use iframe button role.

t-hamano commented on test/e2e/specs/site-editor/navigation.spec.js:
```diff
@@ -0,0 +1,121 @@
+		// Test: Can navigate to a sidebar item and into its subnavigation frame without losing focus
+		// Go to the Pages button
+
+		await editorNavigationUtils.tabToLabel( 'Pages', { times: 10 } );
```
```suggestion
		await editorNavigationUtils.tabToLabel( 'Pages' );
```

I think it is unnecessary here since the default value is defined on the function side

t-hamano commented on test/e2e/specs/site-editor/navigation.spec.js:
```diff
@@ -0,0 +1,121 @@
+		).toBeFocused();
+
+		// Test: Can navigate into the iframe using the keyboard
+		await editorNavigationUtils.tabToLabel( 'Editor Canvas', {
+			times: 10,
+		} );
```
```suggestion
		await editorNavigationUtils.tabToLabel( 'Editor Canvas' );
```

t-hamano commented on test/e2e/specs/site-editor/navigation.spec.js:
```diff
@@ -0,0 +1,121 @@
+		this.pageUtils = pageUtils;
+	}
+
+	async tabToLabel( label, { times = 10 } ) {
```
```suggestion
	async tabToLabel( label, times = 10 ) {
```

t-hamano reviewed (approved):
LGTM ! I left a little comment at the end.

jeryj committed:
Use times param directly instead of as object

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jeryj committed:
Rely on default times value

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jeryj committed:
Rely on default times value

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jeryj commented:
Thanks @t-hamano! I updated based on your excellent feedback.

getdave commented:
I just cherry-picked this PR to the cherry-pick-beta-2 branch to get it included in the next release: 6e5a64f216

ellatrix commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -273,13 +273,16 @@ function Iframe( {
 					// so we need to stop the propagation of this event to avoid
 					// duplication.
-					if (
+					else if (
```
This doesn't make any sense. When you have multiple keydown handlers, all of them should execute

jeryj replied:
The only time the `if` statement above it runs is when a keydown listener is passed to the iframe, which is only in the `view` mode of the site editor. The shortcuts shouldn't be enabled at that point since we're not within the editor.

ellatrix replied:
Ok, but that's not clear from reading this. Also there might be other iframe instances that pass onKeydown and do want shortcuts disabled. Using onKeydown to disable/enable shortcuts seems a bit strange, imo that should be done through a separate prop. Maybe reuse `isPreviewMode`?

jeryj replied:
I'm not sure how all of that should be structured. The only thing important to me in this PR was fixing a bug where you couldn't use the keyboard to enter into Edit mode from View mode. [This onKeydown was not allowed to run](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/block-editor/editor-canvas.js#L64-L71).

Very open to fixing this another way. I wasn't sold on this solution.

ellatrix replied:
No, this is fine, all I'm saying is that the `else` shouldn't have been added?

jeryj replied:
Oh - That's fine with me: https://github.com/WordPress/gutenberg/pull/59474/
